File: _python-archive/devctl/commands/doctor.py
"""`devctl doctor` — diagnose the health of your devctl setup.

Runs a series of independent checks and prints a tidy report.  Exits non-zero
if any check is a hard *error* (so it can gate CI / scripts); *warnings* don't
fail the run.  Like ``run``/``snapshot``, this is registered as a top-level
command rather than a sub-app.

Heavy imports are deferred so importing this module costs only ``import typer``.
"""
from __future__ import annotations

from typing import Literal, NamedTuple

import typer

Status = Literal["ok", "warn", "error"]


class Check(NamedTuple):
    status: Status
    label: str
    detail: str


def _run_checks() -> list[Check]:
    """Collect health checks. Pure (no printing) so it's easy to unit-test."""
    import shutil
    import sys
    from pathlib import Path

    from .. import config, db
    from ..storage import CONFIG_FILE, DEVCTL_DIR, DOTFILES_DIR

    checks: list[Check] = []

    # ── Python ────────────────────────────────────────────────────────────────
    v = sys.version_info
    if v >= (3, 9):
        checks.append(Check("ok", "Python", f"{v.major}.{v.minor}.{v.micro}"))
    else:
        checks.append(Check("error", "Python", f"{v.major}.{v.minor} (need ≥ 3.9)"))

    # ── git (needed for `sync`) ───────────────────────────────────────────────
    git_path = shutil.which("git")
    if git_path:
        checks.append(Check("ok", "git", git_path))
    else:
        checks.append(Check("warn", "git", "not found — `devctl sync` will not work"))

    # ── devctl home is writable ───────────────────────────────────────────────
    try:
        DEVCTL_DIR.mkdir(parents=True, exist_ok=True)
        probe = DEVCTL_DIR / ".devctl-write-probe"
        probe.write_text("ok")
        probe.unlink()
        checks.append(Check("ok", "devctl home", str(DEVCTL_DIR)))
    except OSError as e:
        checks.append(Check("error", "devctl home", f"not writable: {e}"))

    # ── config parses ─────────────────────────────────────────────────────────
    try:
        config.load()
        where = str(CONFIG_FILE) if CONFIG_FILE.exists() else "defaults (no file yet)"
        checks.append(Check("ok", "config", where))
    except Exception as e:  # noqa: BLE001 — report any parse failure as an error
        checks.append(Check("error", "config", f"failed to load: {e}"))

    # ── project registry + stale paths ────────────────────────────────────────
    try:
        projects = db.list_projects()
    except Exception as e:  # noqa: BLE001
        checks.append(Check("error", "registry", f"unreadable: {e}"))
        projects = []
    else:
        stale = [p["name"] for p in projects if not Path(p["path"]).exists()]
        if not projects:
            checks.append(Check("ok", "registry", "no projects registered yet"))
        elif stale:
            shown = ", ".join(stale[:5]) + (" …" if len(stale) > 5 else "")
            checks.append(Check(
                "warn", "registry",
                f"{len(projects)} project(s); {len(stale)} stale path(s): {shown}",
            ))
        else:
            checks.append(Check("ok", "registry", f"{len(projects)} project(s), all paths exist"))

    # ── dotfiles repo (only if sync was initialised) ──────────────────────────
    repo_url = config.load().get("sync", {}).get("repo", "")
    if repo_url:
        if (DOTFILES_DIR / ".git").exists():
            checks.append(Check("ok", "dotfiles repo", str(DOTFILES_DIR)))
        else:
            checks.append(Check(
                "warn", "dotfiles repo",
                "configured but mirror missing — run `devctl sync init`",
            ))

    return checks


def doctor() -> None:
    """Diagnose your devctl setup and report any problems."""
    from rich.console import Console
    from rich.table import Table

    console = Console()
    checks = _run_checks()

    glyph = {"ok": "[green]✓[/]", "warn": "[yellow]![/]", "error": "[red]✗[/]"}
    table = Table(title="devctl doctor", show_lines=False)
    table.add_column("", justify="center")
    table.add_column("Check", style="cyan")
    table.add_column("Detail", style="white")
    for c in checks:
        table.add_row(glyph[c.status], c.label, c.detail)
    console.print(table)

    errors = sum(1 for c in checks if c.status == "error")
    warns = sum(1 for c in checks if c.status == "warn")
    if errors:
        console.print(f"[red]✗ {errors} error(s)[/]" + (f", [yellow]{warns} warning(s)[/]" if warns else ""))
        raise typer.Exit(1)
    if warns:
        console.print(f"[yellow]! {warns} warning(s)[/] — everything else looks good.")
    else:
        console.print("[green]✓ All checks passed.[/]")

File: src/commands/doctor.rs
//! `devctl doctor` — diagnose the health of your devctl setup.
//!
//! Runs independent checks and prints a tidy report. Exits non-zero if any
//! check is a hard *error* (so it can gate CI/scripts); *warnings* don't fail.

use std::path::Path;

use anyhow::Result;
use comfy_table::{presets::UTF8_FULL, Cell, Table};

use crate::storage;
use crate::ui;
use crate::util;
use crate::{config, db};

#[derive(Clone, Copy, PartialEq)]
enum Status {
    Ok,
    Warn,
    Error,
}

struct Check {
    status: Status,
    label: String,
    detail: String,
}

impl Check {
    fn new(status: Status, label: &str, detail: impl Into<String>) -> Self {
        Check {
            status,
            label: label.to_string(),
            detail: detail.into(),
        }
    }
}

/// Collect health checks. Pure (no printing) so the logic is easy to follow.
fn run_checks() -> Vec<Check> {
    let mut checks: Vec<Check> = Vec::new();

    // ── devctl version ──────────────────────────────────────────────────────
    checks.push(Check::new(Status::Ok, "devctl", env!("CARGO_PKG_VERSION")));

    // ── git (needed for `sync`) ─────────────────────────────────────────────
    match util::which("git") {
        Some(p) => checks.push(Check::new(Status::Ok, "git", p.display().to_string())),
        None => checks.push(Check::new(
            Status::Warn,
            "git",
            "not found — `devctl sync` will not work",
        )),
    }

    // ── devctl home is writable ─────────────────────────────────────────────
    let dir = storage::devctl_dir();
    match write_probe(&dir) {
        Ok(()) => checks.push(Check::new(
            Status::Ok,
            "devctl home",
            dir.display().to_string(),
        )),
        Err(e) => checks.push(Check::new(
            Status::Error,
            "devctl home",
            format!("not writable: {e}"),
        )),
    }

    // ── config parses ───────────────────────────────────────────────────────
    match config::load() {
        Ok(_) => {
            let where_ = if storage::config_file().exists() {
                storage::config_file().display().to_string()
            } else {
                "defaults (no file yet)".to_string()
            };
            checks.push(Check::new(Status::Ok, "config", where_));
        }
        Err(e) => checks.push(Check::new(
            Status::Error,
            "config",
            format!("failed to load: {e}"),
        )),
    }

    // ── project registry + stale paths ──────────────────────────────────────
    match db::list_projects() {
        Err(e) => checks.push(Check::new(
            Status::Error,
            "registry",
            format!("unreadable: {e}"),
        )),
        Ok(projects) => {
            let stale: Vec<&str> = projects
                .iter()
                .filter(|p| !Path::new(&p.path).exists())
                .map(|p| p.name.as_str())
                .collect();
            if projects.is_empty() {
                checks.push(Check::new(
                    Status::Ok,
                    "registry",
                    "no projects registered yet",
                ));
            } else if !stale.is_empty() {
                let mut shown = stale.iter().take(5).copied().collect::<Vec<_>>().join(", ");
                if stale.len() > 5 {
                    shown.push_str(" …");
                }
                checks.push(Check::new(
                    Status::Warn,
                    "registry",
                    format!(
                        "{} project(s); {} stale path(s): {}",
                        projects.len(),
                        stale.len(),
                        shown
                    ),
                ));
            } else {
                checks.push(Check::new(
                    Status::Ok,
                    "registry",
                    format!("{} project(s), all paths exist", projects.len()),
                ));
            }
        }
    }

    // ── dotfiles repo (only if sync was initialised) ────────────────────────
    if let Ok(cfg) = config::load() {
        if !cfg.sync.repo.is_empty() {
            if storage::dotfiles_dir().join(".git").exists() {
                checks.push(Check::new(
                    Status::Ok,
                    "dotfiles repo",
                    storage::dotfiles_dir().display().to_string(),
                ));
            } else {
                checks.push(Check::new(
                    Status::Warn,
                    "dotfiles repo",
                    "configured but mirror missing — run `devctl sync init`",
                ));
            }
        }
    }

    checks
}

fn write_probe(dir: &Path) -> Result<()> {
    std::fs::create_dir_all(dir)?;
    let probe = dir.join(".devctl-write-probe");
    std::fs::write(&probe, "ok")?;
    std::fs::remove_file(&probe)?;
    Ok(())
}

pub fn doctor() -> Result<i32> {
    let checks = run_checks();

    let mut table = Table::new();
    table.load_preset(UTF8_FULL);
    table.set_header(vec!["", "Check", "Detail"]);
    for c in &checks {
        let glyph = match c.status {
            Status::Ok => ui::green("✓"),
            Status::Warn => ui::yellow("!"),
            Status::Error => ui::red("✗"),
        };
        table.add_row(vec![
            Cell::new(glyph),
            Cell::new(&c.label),
            Cell::new(&c.detail),
        ]);
    }
    println!("{table}");

    let errors = checks.iter().filter(|c| c.status == Status::Error).count();
    let warns = checks.iter().filter(|c| c.status == Status::Warn).count();
    if errors > 0 {
        let mut msg = ui::red(&format!("✗ {errors} error(s)"));
        if warns > 0 {
            msg.push_str(&format!(", {}", ui::yellow(&format!("{warns} warning(s)"))));
        }
        println!("{msg}");
        return Ok(1);
    }
    if warns > 0 {
        println!(
            "{} — everything else looks good.",
            ui::yellow(&format!("! {warns} warning(s)"))
        );
    } else {
        println!("{}", ui::green("✓ All checks passed."));
    }
    Ok(0)
}
